### PR TITLE
feat: improve dataValidation function

### PR DESCRIPTION
**Pull request by Blakko**
From my tests on a 3k nodes graph, the execution time of this function went from 400ms down to 1ms

### CODE REVIEW

File: packages/core/src/util/validation.ts [hunk 1]
```diff
@@ -43,12 +43,20 @@ export const dataValidation = (data?: GraphData | TreeGraphData): boolean => {
...
 
   // 3. 边的 source 和 target 必须存在于节点 或 Combo中
-  const nodeIds = ((nodes as NodeConfig[]) || []).map(node => node.id);
-  const comboIds = (combos as ComboConfig[])?.map(combo => combo.id);
-  const ids = [...nodeIds, ...comboIds];
-  const nonEdges = ((edges as EdgeConfig[]) || []).find(
-    edge => !ids.includes(edge.source) || !ids.includes(edge.target),
-  );
+  const ids = new Set<string>();
+  if (nodes && (nodes as NodeConfig[]).length) {
```
Comment (Aarebecca) [inline packages/core/src/util/validation.ts:47]: use `nodes?.length` is more concise

**Review by Aarebecca — COMMENTED** [review]
Although it is faster to using `for` than `map` loop, there shouldn't be such a large gap at 3k nodes

**Comment by Blakko** [comment]
The performance gain comes from changing 
`ids.includes(edge.source) || !ids.includes(edge.target)`
into 
`return !ids.has(edge.source) || !ids.has(edge.target);`

I'll test removing the for loops in favor of a `.each` line, I agree there shouldn't be a big difference

**Comment by Blakko** [comment]
Using a forEach to populate the set (rather than the for loop) did show a negligible performance impact, so I made the code tidier

**Comment by Aarebecca** [comment]
It looks like there's something wrong with ci in v4 and this PR will be merged after we fix it

**Pull request merged**